Remove deleted pending workloads from the cache so the visibility API results are correct

astefanutti opened the pull request:
This fixes #1555 and also adds an integration test as suggested in https://github.com/kubernetes-sigs/kueue/issues/1555#issuecomment-1912423532.

#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #1555

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Remove deleted pending workloads from the cache
```

astefanutti commented:
/cc @alculquicondor @mimowo

alculquicondor commented:
/ok-to-test

alculquicondor commented:
Is the issue also reflected in `ClusterQueue.status`? I believe it is. It would be much simpler to test for that and also we could cherry-pick to 0.5.
The visibility server doesn't exist in 0.5, so we can't cherry-pick it.

mimowo commented:
/retest
the failures are due flakes which are now fixed: https://github.com/kubernetes-sigs/kueue/pull/1676

mimowo commented on pkg/visibility/api/rest/utils.go:
```diff
@@ -39,7 +40,7 @@ func newPendingWorkload(wlInfo *workload.Info, positionInLq int32, positionInCq
 		},
 		PositionInClusterQueue: int32(positionInCq),
-		Priority:               *wlInfo.Obj.Spec.Priority,
+		Priority:               priority.Priority(wlInfo.Obj),
```
Is it directly related to the issue or drive-by fix? Why is it needed / useful?

mimowo commented:
> Is the issue also reflected in ClusterQueue.status? I believe it is. It would be much simpler to test for that and also we could cherry-pick to 0.5.

I think the bug is also impacting the `eq.status.pendingWorkloads` counter, so we could use this to verify the fix works at the level of integration tests.

mimowo commented on pkg/visibility/server.go:
```diff
@@ -47,12 +47,11 @@ type server struct {
 
 // CreateAndStartVisibilityServer creates visibility server injecting KueueManager and starts it
 func CreateAndStartVisibilityServer(kueueMgr *queue.Manager, ctx context.Context) {
-	config := newVisibilityServerConfig()
```
Are these changes needed for the fix, or drive-by cleanup? If just cleanup maybe a dedicated PR is preferred IMO.

astefanutti replied:
These changes were necessary for setting up the integration tests for the visibility API, to make it possible to reuse the API server configuration for the test client.

I've removed those integration tests altogether so these changes are not needed anymore.

alculquicondor reviewed (commented):
/lgtm
/approve
/cherry-pick release-0.5

astefanutti commented:
> Is the issue also reflected in `ClusterQueue.status`? I believe it is. It would be much simpler to test for that and also we could cherry-pick to 0.5. The visibility server doesn't exist in 0.5, so we can't cherry-pick it.

You're right! I've removed the integration tests for the visibility API, and added a test case to the existing cluster queue controller integration tests, that verifies the fix.

I've kept it as minimal as possible this time, but feel free to suggest any other test cases that you think could be useful to add there.

mimowo commented on test/integration/controller/core/clusterqueue_controller_test.go:
```diff
@@ -362,6 +362,63 @@ var _ = ginkgo.Describe("ClusterQueue controller", ginkgo.Ordered, ginkgo.Contin
+				gomega.Expect(k8sClient.Create(ctx, workload)).To(gomega.Succeed())
+			})
+
+			// Pending workloads count is incremented as the workload is inadmissible
```
non-blocking nit: I would prefer to turn the comment into ginkgo step. Maybe ginkgo.By("await until the pending workload count is incremented as the workload is inadmissible, because the ResourceFlavors don't exist"). This way the comment is also logged, which can make debugging easier.

astefanutti replied:
Agreed. Actually I saw an existing comment and somehow decided to choose that style, but I would have chosen the `ginkgo.By` stanza otherwise. Let me know if you feel like it's worth I open a follow-up PR.

mimowo replied:
I didn't notice the PR was approved actually, and though you would have time to apply the remark. I'm ok either way, as you feel.

mimowo commented on test/integration/controller/core/clusterqueue_controller_test.go:
```diff
@@ -362,6 +362,63 @@ var _ = ginkgo.Describe("ClusterQueue controller", ginkgo.Ordered, ginkgo.Contin
+				gomega.Expect(k8sClient.Delete(ctx, workload)).To(gomega.Succeed())
+			})
+
+			// Pending workloads count is decrement as the deleted workload has been removed from the queue.
```
non-blocking nit: same here - turn into ginkgo step

mimowo reviewed (commented):
/lgtm
Just non-blocking nits to improve readability of the test.
/assign @tenzen-y @alculquicondor

alculquicondor commented:
/cherry-pick release-0.5
